x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java

```diff
@@ -344,6 +344,9 @@ private AsyncSearchResponse getResponse(boolean restoreResponseHeaders) {
         assert mutableSearchResponse != null;
         checkCancellation();
         AsyncSearchResponse asyncSearchResponse;
+        if (mutableSearchResponse.tryIncRef() == false) {
+            throw new ElasticsearchStatusException("async-search result, no longer available", RestStatus.GONE);
```

Comment:
IIUC the request will still fail, just a different message? (i.e. results in Discover will still not be displayed right?) 

In the race condition you describe we'd have the results on this. Should we not return them from disk instead of failing here?